website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md

```diff
@@ -0,0 +1,412 @@
+# Running gVisor in Production at Scale in Ant
+
+>This post was contributed by [Ant Group](https://www.antgroup.com/), a
+>large-scale digital payment platform. Jianfeng and Yong are engineers at Ant
+>Group working on infrastructure systems, and contributors to gVisor.
+
+At Ant Group, we are committed to keep online transactions safe and efficient.
+Continuously raising the bar of security for potential system-level attacks is
+one of many measures. As a container runtime, gVisor provides container-native
+security without sacrificing resource efficiency. Therefore, it has been on our
+radar since it is open source.
+
+However, there have been performance concerns raised by
+[academia](https://www.usenix.org/system/files/hotcloud19-paper-young.pdf)
+and [industry](https://news.ycombinator.com/item?id=19924036). Users of gVisor
+tend to bear the extra overhead as the tax of security. But we tend to agree
+that [security is no excuse for poor performance](https://sel4.systems/About/seL4-whitepaper.pdf).
+
+In this article, we will present how we identified bottlenecks in gVisor and
+unblocked large-scale production adoption. Our main focus are the CPU
+utilization and latency overhead it brings. Small memory footprint is also a
+valued goal, but not discussed in this blog. As a result of these efforts and
+community improvements, 70% of our applications running on runsc have <1%
+overhead; another 25% have <3% overhead. Some of our mostly valued application
+are the focus of our optimization, and get even better performance compared with
+runc.
+
+The rest of this blog is organized as follows:
+- First, we analyze the cost of different syscall paths in gVisor.
+- Then, a way to profile a whole picture of a instance is proposed to find out
+  if some slow syscall paths are encountered.
+- Some invisible overhead in Go runtime is discussed.
+- At last, a short summary on performance optimization with some other factors
+  on production adoption.
+
+For convenience of discussion, we are targeting KVM-based, or hypervisor-based
+platforms, unless explicitly stated.
+
+## Cost of different syscall paths
+
+[Defense-in-depth](https://gvisor.dev/blog/2019/11/18/gvisor-security-basics-part-1/#defense-in-depth)
```

Comment:
This link can be
../../../../2019/11/18/gvisor-security-basics-part-1/#defense-in-depth